### ISSUE TITLE
Removing Extra functions

### ISSUE THREAD

**Issue by jhkcia** [formulajs/formulajs#79]
some functions do not exist in excel, It is better to remove them and keep our functions the same as excel.

[REGEXEXTRACT, REGEXMATCH, REGEXREPLACE, SPLIT, FINDFIELD, INTERVAL]

Also there are some math formulas which not exist in excel like:
[E, GT, GTE, LT, LTE, EQ, NE, POW, add, minus, divide, multiply]

**Comment by nicolashefti** [comment]
In my opinion these functions are here, do not bother developers much, neither cost time to maintain. Maybe some people are using them.

I would not add new function not existing in Excel. But why removing them?

**Comment by bornova** [comment]
Hi @nicolashefti,
This is an old topic but while I was sorting thru test cases, I did notice we have this `miscellaneous.js` file with functions `FLATTEN`, `ARGS2ARRAY`, `REFERENCE`, `JOIN`, and `NUMBERS`.  None of these functions are used internally by formula.js and they are not Excel function either.  They have test cases associated with them but some of them are confusing.  Take `NUMBERS()`:  The function takes an array of values and returns an array with number values only.  However, the test cases used to test for True/False.  I actually changed the test cases to match the function but I am not even sure of the intended use for this function.

In short, I think it makes sense to remove this `miscellaneous.js` file all together as I believe it is unnecessary.  One less file to maintain, negligible but smaller library size, less confusing code.  What do you think?

**Comment by nicolashefti** [comment]
The doc block speaks for it:

```
/**
 * Formula.js only
 */
```

There are 31 functions with this comment in 7 different files. What about the other functions (outside miscellaneous.js) ? We can also remove them. In this case we can publish a major version and warn about breaking changes.

**Comment by bornova** [comment]
IMHO, they should all be removed to stay true to Excel.  I can open a PR with changes if agreed. Here is a summary of all functions in question:

<html>
<head>
</head>
<body>

Function | File | Action | Excel function? | Used internally?
-- | -- | -- | -- | --
FINDFIELD | database.js | Keep but do not export | No | Yes
INTERVAL | dte-time.js | Remove | No | No
ISBINARY | information.js | Remove | No | No
LN10 | math-trig.js | Remove | No | No
LN2 | math-trig.js | Remove | No | No
LOG10E | math-trig.js | Remove | No | No
LOG2E | math-trig.js | Remove | No | No
E | math-trig.js | Remove | No | No
SQRT1_2 | math-trig.js | Remove | No | No
SQRT2 | math-trig.js | Remove | No | No
ADD | math-trig.js | Remove | No | No
MINUS | math-trig.js | Remove | No | No
DIVIDE | math-trig.js | Remove | No | No
MULTIPLY | math-trig.js | Remove | No | No
GT | math-trig.js | Remove | No | No
GTE | math-trig.js | Remove | No | No
LT | math-trig.js | Remove | No | No
LTE | math-trig.js | Remove | No | No
EQ | math-trig.js | Remove | No | No
NE | math-trig.js | Remove | No | No
POW | math-trig.js | Remove | No | No
REFERENCE | miscellaneous.js | Remove entire file | No | No
JOIN | miscellaneous.js | Remove entire file | No | No
NUMBERS | miscellaneous.js | Remove entire file | No | No
COUNTIN | statistical.js | Remove | No | No
COUNTUNIQUE | statistical.js | Remove | No | No
HTML2TEXT | text.js | Remove | No | No
REGEXEXTRACT | text.js | Remove | No | No
REGEXMATCH | text.js | Remove | No | No
REGEXREPLACE | text.js | Remove | No | No
SPLIT | text.js | Remove | No | No

</body>
</html>

**Comment by nicolashefti** [comment]
Completed with #140

**Comment by mgiorgino** [comment]
Nice cleaning, however, I have some doubts about some removals, in particular the ones corresponding to symbols in excel. For example `ADD`, `MINUS`, `DIVIDE`, `GT`, `GTE` or `LT` are functions that do not exist with these names in Excel, but they do exist as symbols `+`, `-`, `/`, `>`, `>=` or `<`. We could use the js equivalent, but the error management is quite different in js and excel.

I think all functions corresponding to symbols in excel should be restored, maybe to a new module `symbols` and maybe with some util function to get the method from the symbol.

What do you think ?

**Comment by nicolashefti** [comment]
I'm open to discussion and getting the point about error management in the former implementations of `ADD` and other symbol functions.

1) Do you use these functions? If yes, can you tell us a bit more about it? Or is this a hypothetical case?

2) A new user of the library won't be expecting either searching these kinds of functions because they do not exist in Excel (or any spreadsheet software to my knowledge).

@bornova, any thoughts on this?

**Comment by mgiorgino** [comment]
1. Yes we use these functions to evaluate formulas that use as most as possible the same syntax/functions as Excel and provide the same behavior. In our case, this doesn't strictly need to be the same as Excel. But I think some of the functions that were deleted would still be useful (if not necessary) for someone looking to get evaluation of formulas to produce the same result as Excel.

For example, to simply evaluate `= 1 + 2`, we would need a `+` operator. In this case it would match if we use the `+` operator provided by javascript. But if we try to do `1 + NA()`, then it doesn't match anymore, because Excel would display `#N/A` while our js formula would return `'1Error: #N/A'`, that isn't the same as the previous `formulaJs.ADD(1, NA())` that would return the same value as `NA()`, ie an `Error('#N/A')`. And I think errors are not the only difference between Excel `+` and javascript `+`.

So that is why even if `+` from Excel and `+` from javascript look the same, they are not stricly the same, and that is why, IMHO, formulaJs should provide its own `+` and the other functions corresponding to Excel symbols.

2. Maybe a user wouldn't expect them, but he should if he would like to get the same results as in Excel, at least for errors.

**Comment by nicolashefti** [comment]
I understand your points, but I'm wondering if this is the role of Formulajs to address them.

Instead of arguing whether it is good or back, we could create a new module `formulajs/formulajs-extension-symbol` to pack everything that was removed in the #140.

Starting from formulajs backbone it shouldn't be too hard to do.

What do you think? @mgiorgino , @bornova ?

**Comment by mgiorgino** [comment]
I just checked the formulajs website to see if I understood the role of Formulajs correctly and it seems to just raise the same point as me:

https://formulajs.info/#javascript-differs-from-excel

> JavaScript differs from Excel
> 
> Formula.js tries to recreate the formula you know and like from your spreadsheet tool. However the two environments differs from each other.
> 
> For example =2^3 is a valid Excel function that will return 8. The ^ operator means something else in JavaScript (2^3 return 1, you can learn more [here why](https://www.w3schools.com/js/js_bitwise.asp)). With Formula.js you can write formulajs.POWER(2, 3) which is equivalent to Math.pow(2, 3) in plain JavaScript.

Also [this issue](https://github.com/formulajs/formulajs/issues/75) seems to go in my direction, where representation of Excel symbols as functions are useful if we want to evaluate Excel formulas in javascript.

I would be totally satisfied with a separate module, but as I already tried to just re-implement the removed functions on my side, I can tell formulajs is not very "open to extension" for now:

- some util functions are not exposed, making difficult to define our functions in a consistent way with `formulajs`. For example `utils.parseNumber`.
- it is hard to directly use error values (and even impossible for some of them). This would be useful to compare the obtained result, or define our own functions that would return these error values. I, for example, did my best to do it, but it would be simpler for me if it they were exported.

```
import * as formulaJs from "@formulajs/formulajs";

// Retrieve non-exported errors from formulaJs

// Cannot retrieve "nil": never returned...
export const div0 = formulaJs.COT(0);
export const value = formulaJs.DATEVALUE(0);
export const ref = formulaJs.INDEX([], 0, 0);
// Cannot retrieve "name": never returned...
export const num = formulaJs.HOUR("-1.0");
export const na = formulaJs.NA();
export const error = formulaJs.LEN(undefined);
// Cannot retrieve "data": never returned...
```

I would be willing to help.
For extensibility, I would for example suggest to export `utils/common` and `utils/error`.

**Comment by nicolashefti** [comment]
Error objects are already exposed within `formulajs.utils.errors` (see #61 #59). Using a similar technic some of the functions in common.js could be also exposed in `formulajs.utils`.

**Comment by nicolashefti** [comment]
Can you try it would work this way? In this case I can prepare a new repo for the extension.

**Comment by bornova** [comment]
Hi @mgiorgino and @nicolashefti,
I think @mgiorgino may have a good argument.  Although I don't think new users would expect to find these functions it doesn't hurt to have them included I guess.  This was part of the discussion back then and I personally don't have a use for these functions (hence my proposal to remove them). But turns out someone does :)

> Error objects are already exposed within `formulajs.utils.errors` (see #61 #59). Using a similar technic some of the functions in common.js could be also exposed in `formulajs.utils`.

Also looks like the utils are not exported anymore.  After the 3.0 refactor, they seem to be left out unintentionally.

So, I would suggest the following:
- Re export errors as `utils.errors`, as it was introduced in #61
- Introduce a new file named `symbol.js` under `utils` folder and export the functions as `utils.symbols`
- Document the changes in `README.md`

Here is the PR: #147
We can continue the discussion on this implementation there.

**Comment by nicolashefti** [comment]
To quote myself :)

> In my opinion these functions are here, do not bother developers much, neither cost time to maintain. Maybe some people are using them.
> 
> I would not add new function not existing in Excel. But why removing these existing ones?

@bornova to summarize, we re-introduce the Symbol-functions, but all the text / statistical / miscellaneous stay out ? It sounds like a good compromise. No additional extension repo to maintain. Let's keep things simple.

@mgiorgino is that ok to have them under './utils' and not the same root level as the other function?

**Comment by mgiorgino** [comment]
@nicolashefti Yes, thanks. I would just like to have them 😀, so I completely agree with your suggestions.

Hi @bornova, thanks for the PR ! I'll have a look at it, but it seems perfect from your description.

**Comment by bornova** [comment]
> @bornova to summarize, we re-introduce the Symbol-functions, but all the text / statistical / miscellaneous stay out ? It sounds like a good compromise. No additional extension repo to maintain. Let's keep things simple.

I can see the value in keeping these functions for the main math symbols if it helps other users.  And since they are not Excel functions, makes sense to keep them in a separate file. A separate extension repo would be one more repo to maintain like you said.

Let me know your feedback on naming conventions, file name and location, etc.

The way I implemented it in the PR, the usage would be:

```javascript
import * as formulajs from '@formulajs/formulajs'

formulajs.utils.symbols.ADD(1, 2)    // 3
formulajs.utils.errors.nil    // Error: #NULL!
```

**Comment by nicolashefti** [comment]
New release, I hope this discussion is closed for good 😄

**Comment by mgiorgino** [comment]
Yes sorry for the delay. I have been able to successfully use the new release 4.2.0, thank you all !

**Comment by mytharcher** [comment]
Could you help to update the documentation on official site to tell that some function are not available and removed from 4.2.0 (in case any confusion)?

e.g.:

* COUNTUNIQUE
* REGEXMATCH
* SPLIT
* ...

**Comment by 110526** [comment]
> In my opinion these functions are here, do not bother developers much, neither cost time to maintain. Maybe some people are using them.
> 
> I would not add new function not existing in Excel. But why removing these existing ones?

Why did your thoughts waver

**Comment by nicolashefti** [comment]
@110526 you're commenting on an issue closed 1 year ago. If you want to propose something new feel free to open a new thread.
